### ISSUE TITLE
use standard cp arguments

### ISSUE THREAD

**Issue by bradain** [nodejs/node-gyp#571]
The cp command takes an argument -a as a shortcut for -pPR.
While this works on most Linux distributions, it is not available on certain other operating systems such as OpenBSD, while the standard -pPR always works. 
Would you consider replacing cp -af with the functionally equivalent cp -fpPR to improve OS support?

**Comment by drook** [comment]
Same stuff on Solaris.

**Comment by Fishrock123** [comment]
Huh. I thought I commented on this.

`cp` resides in gyp itself here, so I'm not sure if we can do that? cc @TooTallNate, do we still ever get upstream updates from gyp?

Should we just have a nodejs/gyp that both node and node-gyp can use?

**Comment by alexijs** [comment]
Same stuff on Oracle Solaris 11.2

**Comment by TooTallNate** [comment]
@Fishrock123 Historically I've just updated gyp here when node.js/io.js updated its own copy of gyp (probably behind at this point!)

**Comment by alexijs** [comment]
Is there a way to resolve this issue by changing cp flags? Compilation of all packages via node-gyp fails on Oracle Solaris 11.2 because Oracle doesn't support GNU flag -a.

I found a workaround by compiling package locally, creating a package structure, and moving *.node, package.json, and corresponding js files there.
However, it defeats a purpose of autobuilder and too cumbersome because I have to compile package after package.

Also, it looks like running Demeteorizer is not enough and I have to write custom builder that would replace packages from my dev (OS X Maverics) with those locally compiled on Solaris.

Im trying to deploy my Meteor app on Solaris via Demeteorizer and have problems with fibers and bcrypt.

Any help would be appreciated.

**Comment by drook** [comment]
You could also play with Makefile (or, if I don't remember correctly, with any other template issuing "cp -r").

**Comment by bnoordhuis** [comment]
Little known fact but you can override the cp command: `make cmd_copy='ln -f $< $@ || cp -PRfp $< $@'`

If someone sends a patch to the upstream gyp project, I'll be happy to land it here (assuming it's accepted upstream, of course.)  Grep for `copy_archive_args` in pylib/gyp/generator/make.py.

**Comment by mhdawson** [comment]
As an FYI we just did a similar change for AIX here https://codereview.chromium.org/1368133002/

**Comment by bnoordhuis** [comment]
@mhdawson You can make that part of #753 if you want.

**Comment by GeorgeBailey** [comment]
> @GeorgeBailey: I wonder if you experience the same problems on your system.

@alexijs, Not thus far. Perhaps it is because all the `npm` modules I tried so far were JS-only. (I hope I correctly understand the purpose of `node-gyp`)

**Comment by Fishrock123** [comment]
See https://github.com/nodejs/node/pull/3487

**Comment by mhdawson** [comment]
FYI - Imran is working on a PR  for the same update to gyp as in nodejs/node#3487 but for node-gyp. Should be submitted soon.

**Comment by mhdawson** [comment]
The PR mentioned above https://github.com/nodejs/node-gyp/pull/781

**Comment by TomFreudenberg** [comment]
On OpenBSD 5.7 I patched the files:

(changing `cp -af` into `cp -pPRf`)

```
node-gyp/0003-gyp-don-t-use-links-at-all-just-copy-the-files-inste.patch:+cmd_copy = rm -rf "$@" && cp -pPRf "$<" "$@"
node-gyp/0003-gyp-don-t-use-links-at-all-just-copy-the-files-inste.patch:+      command='rm -rf $out && cp -pPRf $in $out')
node-gyp/gyp/pylib/gyp/generator/make.py:cmd_copy = rm -rf "$@" && cp -pPRf "$<" "$@"
node-gyp/gyp/pylib/gyp/generator/ninja.py:      command='rm -rf $out && cp -pPRf $in $out')
```

That works successfully

It would be great if the PR would be applied to reduce the pain on OpenBSD and others.

Thanks, Tom

**Comment by bnoordhuis** [comment]
I filed https://codereview.chromium.org/1483473002/ to fix `cp` on OpenBSD.

**Comment by bnoordhuis** [comment]
Fixed (on OpenBSD anyway) in v3.2.1.
